File: DynamicProgramming/AdjacentBitCounts/count.py
class Solution:
    def countStrings(self, n, k):
        MOD = 1000000007

        dp = [[[-1 for _ in range(2)] for _ in range(k + 1)] 
              for _ in range(n + 1)]

        def doDP(n, k, lastChar):
            if n == 0:
                return 1 if k == 0 else 0

            if k < 0:
                return 0

            if dp[n][k][lastChar] != -1:
                return dp[n][k][lastChar]

            res = 0

            if lastChar:
                res = doDP(n - 1, k, 0) % MOD
                res = (res + doDP(n - 1, k - 1, 1)) % MOD
            else:
                res = doDP(n - 1, k, 0) % MOD
                res = (res + doDP(n - 1, k, 1)) % MOD

            dp[n][k][lastChar] = res
            return res

        return (doDP(n - 1, k, 0) +
                doDP(n - 1, k, 1)) % MOD

File: DynamicProgramming/AdjacentBitCounts/count.cpp
#include <cstring>

class Solution {
  public:
    int MOD = 1000000007;
    int dp[1005][1005][2];
    int doDP(int n, int k, int lastChar){
        if(n==0)return k==0;
        else if(k<0)return 0;
        else if(dp[n][k][lastChar]!=-1)return dp[n][k][lastChar];
        int res = 0;
        if(lastChar){
            res = (doDP(n-1,k,0)%MOD + res%MOD)%MOD;
            res = (doDP(n-1,k-1,1)%MOD + res%MOD)%MOD;
        }else{
            res = (doDP(n-1,k,0)%MOD + res%MOD)%MOD;
            res = (doDP(n-1,k,1)%MOD + res%MOD)%MOD;
        }
        return dp[n][k][lastChar] = res;
    }
    int countStrings(int n, int k) {
        // code here
        memset(dp,-1,sizeof(dp));
        return (doDP(n-1,k,0)%MOD + doDP(n-1,k,1)%MOD)%MOD;
    }
};